TestcaseJsStr should ignore unmatched quote characters while scanning

When reducing a testcase with TestcaseJsStr containing broken strings, this assertion is hit:

    AssertionError: Unexpected EOF looking for end of string (')

A better approach would be to continue parsing as though the starting string were not there. Eg.

    'abc', "'123'

Currently this hits the above assertion. I propose to parse it as follows (ignoring the unmatched `"`):

    '
    a
    b
    c
    ', "'
    1
    2
    3
    '

--- a/src/lithium/reducer.py
+++ b/src/lithium/reducer.py
@@ -162,14 +162,15 @@ class TestcaseJsStr(TestcaseChar):
         # if we hit EOF while looking for end of string, we need to rewind to the state before we matched on
         # that quote character and try again.
         while self._instr is not None:
-            for idx in reversed(range(len(self.parts))):
+            idx = None
+            for idx in reversed(range(len(self.parts))):  # pylint: disable=range-builtin-not-iterating
                 if self.parts[idx].endswith(self._instr) and idx not in self._chars:
                     break
             else:
                 raise RuntimeError("error while backtracking from unmatched %s" % (self._instr,))
             self.parts, rest = self.parts[:idx+1], b"".join(self.parts[idx+1:])
-            self._chars = [c for c in self._chars if c < idx]
-            self._instr = None
+            self._chars = [c for c in self._chars if c < idx]  # pylint: disable=attribute-defined-outside-init
+            self._instr = None  # pylint: disable=attribute-defined-outside-init
             self.readTestcaseLine(rest)
         del self._instr
 

--- a/src/lithium/reducer.py
+++ b/src/lithium/reducer.py
@@ -159,7 +159,18 @@ class TestcaseJsStr(TestcaseChar):
 
         super(TestcaseJsStr, self).readTestcase(filename)
 
-        assert self._instr is None, "Unexpected EOF looking for end of string (%s)" % self._instr
+        # if we hit EOF while looking for end of string, we need to rewind to the state before we matched on
+        # that quote character and try again.
+        while self._instr is not None:
+            for idx in reversed(range(len(self.parts))):
+                if self.parts[idx].endswith(self._instr) and idx not in self._chars:
+                    break
+            else:
+                raise RuntimeError("error while backtracking from unmatched %s" % (self._instr,))
+            self.parts, rest = self.parts[:idx+1], b"".join(self.parts[idx+1:])
+            self._chars = [c for c in self._chars if c < idx]
+            self._instr = None
+            self.readTestcaseLine(rest)
         del self._instr
 
         # self._chars is a list of all the indices in self.parts which are chars

--- a/tests/test_lithium.py
+++ b/tests/test_lithium.py
@@ -810,6 +810,18 @@ class TestcaseTests(TestCase):
         assert t.before == b"'"
         assert t.parts == [b"x", b"a", b"b", b"c", b"x"]
         assert t.after == b"'"
+        with open("a.txt", "wb") as f:
+            f.write(b"'x'abcx'")
+        t.readTestcase("a.txt")
+        assert t.before == b"'"
+        assert t.parts == [b"x"]
+        assert t.after == b"'abcx'"
+        with open("a.txt", "wb") as f:
+            f.write(b"'x\"abc\"x")
+        t.readTestcase("a.txt")
+        assert t.before == b"'x\""
+        assert t.parts == [b"a", b"b", b"c"]
+        assert t.after == b"\"x"
 
     def test_symbol(self):
         t = lithium.TestcaseSymbol()

